src/Microsoft.ML.StandardLearners/Standard/SdcaCatalog.cs

```diff
@@ -31,7 +31,9 @@ public static class SdcaRegressionExtensions
         /// if both are present and have different values.
         /// The columns names, however need to be provided directly, not through the <paramref name="advancedSettings"/>.</param>
         public static SdcaRegressionTrainer StochasticDualCoordinateAscent(this RegressionContext.RegressionTrainers ctx,
-        string label = DefaultColumnNames.Label, string features = DefaultColumnNames.Features, string weights = null,
```

Comment:
> = DefaultColumnNames.Label [](start = 20, length = 27)

why do we remove the defaults?

Reply:
we talked about them being error prone, better ask the user to define them explicitly. 

---
In reply to: [231231911](https://github.com/dotnet/machinelearning/pull/1524#discussion_r231231911) [](ancestors = 231231911)

Reply:
@TomFinley do you have a preference? I would prefer to keep the defaults. Otherwise, why do we even have `DefaultColumnNames` in the API?

---
In reply to: [231234097](https://github.com/dotnet/machinelearning/pull/1524#discussion_r231234097) [](ancestors = 231234097,231231911)

Reply:
I think the default arguments make sense for the case when the user doesn't supply anything, the pipeline just works. This is not the case for the column names; if they are not "Label" and "Features" in the datasets the pipelines fail, sometimes with esoteric messages. 

I fully endorse the default for the numeric values: they guide the user about ranges, and they just work most of the time. 
I am also ok with Default column names for the columns components produce and hard-code like "Score" "PredictedLabel"; but IMO hardcode the inputs is misleading and will hide errors like the ones we have fixed so far (not passing the column names at all, and things still work because our test datasets and the default col names match). 

---
In reply to: [231367619](https://github.com/dotnet/machinelearning/pull/1524#discussion_r231367619) [](ancestors = 231367619,231234097,231231911)

Reply:
I think generally it is a good policy to be cautious w.r.t. expanding defaults, as @sfilipi suggests. Being too aggressive with your defaults can lead to more confusion, not less.

> I think the default arguments make sense for the case when the user doesn't supply anything, the pipeline just works. This is not the case for the column names;

I would argue that it is the case for column names more often than we might think. The text loader when no settings are defined has the default behavior, which has proven to be valuable, that it considers the first field a scalar `Label` and subsequent fields a vector `Features`.

Further, while our text parser is built so that it can accommodate a diverse set of inputs (if properly configured), we should not loose sight of the fact that *generally* file formats for machine learning are far more restrictive than this. Consider formats like [SVM-light](http://svmlight.joachims.org/), the more esoteric Bing formats, (we have loaders for the previous that we haven't yet had the chance to open source) or the [VW format](http://hunch.net/~vw/). In these formats, "label" and "features" are priviledged concepts, and where we've written loaders for them, we name these things the columns `Label` and `Features`. Or the "wizard" case, where we apply heusitics to figure out how to load data, what else will the result be named but `Label` and `Features`?

Now consider two populations of people, under two conditions.

Condition A, with defaults!

1. Someone uses SVM-light loader, or the Bing BIN loader, or the text loader under default settings. (Or, in future, a hypothetical VW loader.) These will provide two columns `Label` and `Features`. With defaults, they call `Read()` then `Fit()`. Things work.

2. Now, you are talking about someone that overrides the default behavior of `TextLoader` (again, that default behavior ) and provides explicit names for their columns. They call `Fit` on this trainer. Things *fail* with the message that it could not find a label or feature column named `Label` or `Features`.

Condition B, without defaults!

1. Same actions, but nwo things fail. They need to provide these column names, but how could they reasonably know what the column names are? Indeed, someone just coming to ML.NET for the first time and just fires up a session with SVM-light loader for the first time, how could we expect them to even know what a column *is*?

2. Same action, things *still* fail.

It is far from clear to me that A2 is significantly more confusing than B2 -- in order to hit B2 we need a user that is both (1) smart enough to figure out how to come up with names for columns but (2) not smart enough to figure out that things downstream might want those names. Even if such a person exists, the situation of A2 is recoverable.

We must, I think, all agree that B1 is borderline disastrously confusing. And A1 both work.

*Some* people have really confusing pipelines with multiple featurizers, custom data, and all this stuff. But some people just have data, already nicely structured, and just want to read and train. IDataView and Columns are sort of one of those concepts you need to know if you're doing complex featurization, but if you don't, you don't really need to know it, and I'd like to keep it that way if we can.

Therefore, my own sense is that @Zruty0 is correct, and we ought to have defaults.

---
In reply to: [231388508](https://github.com/dotnet/machinelearning/pull/1524#discussion_r231388508) [](ancestors = 231388508,231367619,231234097,231231911)

Reply:
Thanks for taking a side, @TomFinley, appreciate the level of detail. My observation is that overriding the behavior of the TextLoader is actually the norm, and unless your immediate component is a trainerEstimator, you need the names of the columns to transform them further. 


Examples of ML.Net issues with non-standard TextLoader:
https://github.com/dotnet/machinelearning/issues/420
https://stackoverflow.com/questions/52689363/adding-data-from-new-textloader-into-exisitng-pipeline-ver-0-6 
https://elbruno.com/2018/06/19/mlnet-loading-data-in-our-learningpipeline-with-list-lists-for-ever/
https://carlos.mendible.com/2018/06/10/simple-machine-learning-with-dotnet-core-sample/

Reply:
I wrote more about my experience with the default column names in #1360.